src/wp-includes/media.php

```diff
@@ -5640,9 +5693,11 @@ function wp_get_loading_optimization_attributes( $tag_name, $attr, $context ) {
 		return $loading_attrs;
 	}
 
+	$maybe_size_less_attrs = null;
+
 	// For any resources, width and height must be provided, to avoid layout shifts.
 	if ( ! isset( $attr['width'], $attr['height'] ) ) {
-		return $loading_attrs;
```

Comment:
@felixarntz 

This is too early to bail, as images without these attributes should still get the `decoding` attribute.

Reply:
It is indeed too early to return for `decoding`, but we can fix that in a more elegant way without introducing a new variable like `$maybe_size_less_attrs`, which is hard to follow. Here's how:
* Change the order of the early return here (dimension attributes check) and the early return below based on the context (the `doing_filter( 'the_content' )` part).
* Then add `decoding="async"` between those two conditions: That way the function still returns early in case of those contexts, but applies the `decoding` attribute prior to the early return due to missing dimension attributes.